Don't error our on deprecation with upstream compiler

mihails-strasuns-sociomantic opened the pull request:
This is necessary to be able to test ocean as part of upstream
regression test suite.

mihails-strasuns-sociomantic committed:
Move part of the build configuration to Config.mak

It is recommended by makd to use dedicated configuration file when
amount of options is more than few
(https://github.com/sociomantic-tsunami/makd#config-mak). Ocean build
configuration is rather convoluted now so it will make actual Build.mak
more readable to split it.

Only remaining bits are the one depending on DVER which is set after
Config.mak

nemanja-boric-sociomantic commented:
Was moving the stuff from `Build.mak` to `Config.mak` necessary? The commit message could explain the reason.

mihails-strasuns-sociomantic commented:
> Was moving the stuff from Build.mak to Config.mak necessary?

No, I did it only because I have noticed it. Reason is that it is recommended by MakD documentation :P

nemanja-boric-sociomantic commented:
Fair enough. Just mention it in the commit message (as it will surely confuse the many) and LGTM.

mathias-lang-sociomantic commented:
Exactly what Nemanja said, add a message to the first commit and LGTM.

leandro-lucarella-sociomantic commented on Config.mak:
```diff
@@ -3,11 +3,15 @@ export ASSERT_ON_STOMPING_PREVENTION=1
-else
+ifeq ($(DVER),2)
 # Open source Makd uses dmd by default
 DC ?= dmd
```
I don't get this, if we are in D2, then assign the same as makd uses by default?

nemanja-boric-sociomantic replied:
Ping @mihails-strasuns-sociomantic (I know it's not your fault ;-) ). This is the left over when we had `DC ?= dmd-transitional` here IIRC.

leandro-lucarella-sociomantic replied:
Ah, right, it was just moved in this commit.

leandro-lucarella-sociomantic reviewed (changes requested):
One comment, plus what @nemanja-boric-sociomantic said, plus please also add more clarifications on the other commit message saying exactly why is this needed for upstream test. I don't really get what's going on with these changes or why they are necessary.

leandro-lucarella-sociomantic commented:
Like "if DC is blah, we do bleh to avoid X doing Y" or something on those lines.

mihails-strasuns-sociomantic committed:
Clean up and add explanations to Config.mak

mihails-strasuns-sociomantic committed:
Rework how deprecation flag is being deduced

- In D1 it is implemented by filtering away `-di` flag from default flags
- In D2 it is enabled by simply adding `-de` to compiler flags
- If "ALLOW_DEPRECATIONS" flag is set, both above cases are ignored

mihails-strasuns-sociomantic commented:
Updated with more descriptive commits

mihails-strasuns-sociomantic commented:
@leandro-lucarella-sociomantic if we can get this in today, I will be able to include it in v4.0.0-alpha.1 and start testing upstream ^_^

nemanja-boric-sociomantic commented on Config.mak:
```diff
@@ -1,17 +1,19 @@
 override DFLAGS := $(filter-out -di,$(DFLAGS)) -v2 -v2=-static-arr-params -v2=-volatile
 else
-# Open source Makd uses dmd by default
-DC ?= dmd
```
Ah, you sorted it out.

leandro-lucarella-sociomantic commented on Config.mak:
```diff
@@ -0,0 +1,17 @@
+
+override DFLAGS += -w -version=GLIBC
+
+ifeq ($(DVER),1)
```
Something to have in mind here, this will not do what you want if `DVER` is not passed explicitly to `make`, because `DVER` is defined **after** `Config.mak` was read (look at the docs, it says there are only a few variables available to `Config.mak`). I think this should be kept in `Build.mak`.

leandro-lucarella-sociomantic commented on Config.mak:
```diff
@@ -0,0 +1,17 @@
+override DFLAGS += -w -version=GLIBC
+
+ifeq ($(DVER),1)
+override DFLAGS := $(filter-out -di,$(DFLAGS)) -v2 -v2=-static-arr-params -v2=-volatile
```
Same for `DFLAGS`, you can define the defaults in here, but you can't `filter-out` because it's not defined at this stage.

leandro-lucarella-sociomantic commented on Config.mak:
```diff
@@ -6,12 +6,15 @@ export ASSERT_ON_STOMPING_PREVENTION=1
-override DFLAGS := $(filter-out -di,$(DFLAGS)) -v2 -v2=-static-arr-params -v2=-volatile
-else
+# symbols internally. Only exception is upstream D2 compiler which may introduce
+# new language deprecations ocean was not yet adjusted to.
```
This also seems like a very niche case (the one testing the development compiler). What about adding some option `DMD_DEVEL` or something, to indicate we are testing with a in-development compiler, because if we would like to test with a final and stable upstream dmd, we would want to error on deprecations, right?

mihails-strasuns-sociomantic replied:
I don't expect it matter in practice, but sounds reasonable.

leandro-lucarella-sociomantic commented on Config.mak:
```diff
@@ -6,12 +6,15 @@ export ASSERT_ON_STOMPING_PREVENTION=1
+ifeq ($(DC),dmd-transitional)
 override DFLAGS += -de
 endif
+else
```
Maybe introduce some indentation to make more clear which `else` belongs to each `ifeq`?

leandro-lucarella-sociomantic commented:
> I will be able to include it in v4.0.0-alpha.1

Can we **please** use `-alpha1` so it's consistent with the `-rcX` releases? :innocent:

mihails-strasuns-sociomantic committed:
Remove confusing comment and explicit DC assignment

In fact current makd defaults define same value already

mihails-strasuns-sociomantic committed:
Remove D1 -v2 flags

With current stage of ocean D2 migration those serve no purpose and only
make Build.mak more confusing

mihails-strasuns-sociomantic commented:
Updated.

mihails-strasuns-sociomantic commented:
> Can we please use -alpha1 so it's consistent with the -rcX releases?

For the reference to observers - https://github.com/sociomantic-tsunami/neptune/issues/132 :)

leandro-lucarella-sociomantic commented on Build.mak:
```diff
@@ -1,17 +1,3 @@
-# This will make D2 unittests fail if stomping prevention is triggered
```
Typo in commit message "Move port ..." should be "part" I guess.

leandro-lucarella-sociomantic commented on Build.mak:
```diff
@@ -1,8 +1,6 @@
 override DFLAGS := $(filter-out -di,$(DFLAGS)) -v2 -v2=-static-arr-params -v2=-volatile
 else
-# Open source Makd uses dmd by default
-DC ?= dmd
```
Typo in commit message "confusin" -> "confusing", maybe add that it's removed because is the makd default.

leandro-lucarella-sociomantic commented on Build.mak:
```diff
@@ -1,9 +1,12 @@
-override DFLAGS += -de
+# symbols internally. Disable it on explicit flag to make possible regression
+# testing in D upstream.
+ifndef ALLOW_DEPRECATIONS
```
I won't say this is a blocker, but I would compare against `ALLOW_DEPRECATIONS=1` to avoid confusion if someone thinks it can explicitly disable deprecations by using `ALLOW_DEPRECATIONS=0`, which is a common practice in Makd.

leandro-lucarella-sociomantic reviewed (approved):
Minus minor comments, look good to me, feel free to self-merge once the typos are fixed!
